File: app/src/main/java/com/fdi/pad/pethouse/activity_login.java
package com.fdi.pad.pethouse;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;

/**
 * Actividad que define la pantalla inicial de login.
 */
public class activity_login extends AppCompatActivity implements View.OnClickListener {
    /*------------------------------ATRIBUTOS----------------------------*/
    /**
     * Etiqueta para los logs de inicio de sesión.
     */
    private static final String TAG = "Login";
    /**
     * Botón para iniciar el registro de un usuario.
     */
    private Button button_register;
    /**
     * Botón para iniciar la sesión del usuario.
     */
    private Button button_enter;
    /**
     * Casillero de texto donde se encuentra el email del usuario.
     */
    private EditText edit_text_email;
    /**
     * Casillero de texto donde se encuentra la contraseña del usuario.
     */
    private EditText edit_text_password;
    /**
     * Autencicador de la aplicación dado por la tecnología FireBase.
     */
    private FirebaseAuth my_authentication;

    /*--------------------------ETAPAS---------------------------------*/
    /**
     * Creación de la actividad.
     * @param savedInstanceState El estado de la aplicación en un paquete.
     */
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        my_authentication = FirebaseAuth.getInstance();

        button_register = (Button) findViewById(R.id.buttonRegisterLogin);
        button_register.setOnClickListener(this);
        button_enter = (Button) findViewById(R.id.buttonEnterLogin);
        button_enter.setOnClickListener(this);

        edit_text_email = (EditText) findViewById(R.id.editTextEmailLogin);
        edit_text_password = (EditText) findViewById(R.id.editTextPasswordLogin);
    }

    /**
     * La actividad comienza a ejecutarse.
     */
    @Override
    public void onStart() {
        super.onStart();
        /*Comprobamos si hay algún usuario que ya ha iniciado la sesión.*/
        FirebaseUser current_user = my_authentication.getCurrentUser();
        /*En caso de que se haya iniciado, se procede a mostrar la pantalla principal.*/
        if (current_user != null) {
            login(current_user);
        }
    }

    /*--------------------------EVENTOS---------------------------------*/
    /**
     * Ejecuta un procedimiento al tocar un elemento de la actividad.
     * @param v Elemento que ha sido tocado.
     */
    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.buttonRegisterLogin:
                RegisterButton();
                break;
            case R.id.buttonEnterLogin:
                EnterButton();
                break;
        }
    }

    /*--------------------------MÉTODOS PRIVADOS---------------------------------*/
    /**
     * Procede a cambiar a la actividad "home".
     * @param current_user Usuario logueado.
     */
    private void login(FirebaseUser current_user) {
        Intent intent = new Intent(activity_login.this, activity_home.class);
        startActivity(intent);
    }
    /**
     * Comienza el flujo del registro de un usuario.
     */
    private void RegisterButton() {
        Intent intent = new Intent(activity_login.this, regNombre.class);
        startActivity(intent);
    }
    /**
     * Si los valores son válidos, loguea al usuario.
     */
    private void EnterButton() {
        /*Cogemos los valores de los casilleros.*/
        String email = edit_text_email.getText().toString();
        String password = edit_text_password.getText().toString();
        if (!validateForm(email, password)) {
            return;
        }
        /*Ejecutamos la autentificación del usuario.*/
        my_authentication.signInWithEmailAndPassword(email, password)
                .addOnCompleteListener(this, new OnCompleteListener<AuthResult>() {
                    @Override
                    public void onComplete(@NonNull Task<AuthResult> task) {
                        if (task.isSuccessful()) {
                            // Sign in success, update UI with the signed-in user's information
                            Log.d(TAG, "signInWithEmail:success");
                            FirebaseUser user = my_authentication.getCurrentUser();
                            login(user);
                        } else {
                            // If sign in fails, display a message to the user.
                            Log.w(TAG, "signInWithEmail:failure", task.getException());
                            Toast.makeText(activity_login.this, "Autentificación fallida.",
                                    Toast.LENGTH_SHORT).show();
                        }
                    }
                });
    }

    /**
     * Comprueba que los casilleros no estén vacíos.
     * @return Casilleros correctos
     */
    private boolean validateForm(String email, String password) {
        boolean correct = true;

        if (TextUtils.isEmpty(email)) {
            edit_text_email.setError("Requerido.");
            correct = false;
        } else {
            edit_text_email.setError(null);
        }

        if (TextUtils.isEmpty(password)) {
            edit_text_password.setError("Requerido.");
            correct = false;
        } else {
            edit_text_password.setError(null);
        }
        return correct;
    }
}

File: app/src/main/java/com/fdi/pad/pethouse/ActivityLogin.kt
package com.fdi.pad.pethouse

import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.text.TextUtils
import android.util.Log
import android.util.Patterns
import android.view.View
import android.widget.Toast
import com.fdi.pad.pethouse.home.ActivityHome

import com.fdi.pad.pethouse.userRegistration.ActivityRegistrationName
import com.google.firebase.auth.FirebaseAuth
import kotlinx.android.synthetic.main.activity_login.*

/**
 * Actividad que define la pantalla inicial de login.
 */
class ActivityLogin : AppCompatActivity() {

    //region ATRIBUTOS

    /**
     * Etiqueta para los logs de inicio de sesión.
     */
    private val tag = "Login"

    /**
     * Autentificador de la aplicación dado por la tecnología FireBase.
     */
    private var authentication: FirebaseAuth? = null

    //endregion

    //region EVENTOS

    /**
     * Creación de la actividad.
     *
     * @param savedInstanceState El estado de la aplicación en un paquete.
     */
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)

        authentication = FirebaseAuth.getInstance()

        buttonRegister.setOnClickListener { registerButton() }
        buttonEnter.setOnClickListener { enterButton() }
    }

    /**
     * La actividad comienza a ejecutarse.
     */
    public override fun onStart() {
        super.onStart()
        /*Comprobamos si hay algún usuario que ya ha iniciado la sesión.*/
        val currentUser = authentication!!.currentUser
        /*En caso de que se haya iniciado, se procede a mostrar la pantalla principal.*/
        if (currentUser != null) {
            login()
        }
    }

    /**
     * La actividad se reestablece
     */
    override fun onRestart() {
        super.onRestart()
        editTextEmail.text.clear()
        editTextPassword.text.clear()
    }

    //endregion

    //region MÉTODOS PRIVADOS

    /**
     * Procede a cambiar a la actividad "home".
     */
    private fun login() {
        progressBarLogin.visibility = View.INVISIBLE
        val intent = Intent(this@ActivityLogin, ActivityHome::class.java)
        startActivity(intent)
    }

    /**
     * Comienza el flujo del registro de un usuario.
     */
    private fun registerButton() {
        val intent = Intent(this@ActivityLogin, ActivityRegistrationName::class.java)
        startActivity(intent)
    }

    /**
     * Si los valores son válidos, loguea al usuario.
     */
    private fun enterButton() {
        /*Cogemos los valores de los casilleros.*/
        val email = editTextEmail.text.toString()
        val password = editTextPassword.text.toString()
        if (!validateForm(email, password)) {
            return
        }
        progressBarLogin.visibility = View.VISIBLE
        /*Ejecutamos la autentificación del usuario.*/
        authentication!!.signInWithEmailAndPassword(email, password)
                .addOnCompleteListener(this) { task ->
                    if (task.isSuccessful) {
                        // Sign in success, update UI with the signed-in user's information
                        Log.d(tag, "signInWithEmail:success")
                        login()
                    } else {
                        // If sign in fails, display a message to the user.
                        Log.w(tag, "signInWithEmail:failure", task.exception)
                        Toast.makeText(this@ActivityLogin, "Autentificación fallida.",
                                Toast.LENGTH_SHORT).show()
                        progressBarLogin.visibility = View.INVISIBLE
                    }
                }
    }

    /**
     * Comprueba que los casilleros no estén vacíos.
     *
     * @return Casilleros correctos
     */
    private fun validateForm(email: String, password: String): Boolean {
        var correctEmail = true
        var correctPassword = true

        when {
            TextUtils.isEmpty(email) -> {
                editTextEmail.error = getString(R.string.required_field)
                correctEmail = false
            }
            Patterns.EMAIL_ADDRESS.matcher(email).matches() -> editTextEmail.error = null
            else -> {
                editTextEmail.error = getString(R.string.incorrect_email_format)
                correctEmail = false
            }
        }

        if (TextUtils.isEmpty(password)) {
            editTextPassword.error = getString(R.string.required_field)
            correctPassword = false
        } else {
            editTextPassword.error = null
        }
        return correctEmail && correctPassword
    }

    //endregion

}

